modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/aware/SegmentArchiveSizeStorage.java

```diff
@@ -17,65 +17,132 @@
 import org.apache.ignite.internal.IgniteInterruptedCheckedException;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.configuration.DataStorageConfiguration.UNLIMITED_WAL_ARCHIVE;
 
 /**
  * Storage WAL archive size.
  * Allows to track the exceeding of the maximum archive size.
  */
 class SegmentArchiveSizeStorage {
-    /** Current WAL archive size in bytes. */
+    /** Current WAL archive size in bytes. Guarded by {@code this}. */
     private long curr;
 
-    /** Reserved WAL archive size in bytes. */
-    private long reserved;
+    /** Flag of interrupt waiting on this object. Guarded by {@code this}. */
+    private boolean interrupted;
+
+    /** Minimum size of the WAL archive in bytes. */
+    private final long minWalArchiveSize;
+
+    /** Maximum size of the WAL archive in bytes. */
+    private final long maxWalArchiveSize;
+
+    /**
+     * Segment sizes. Mapping: segment idx -> size in bytes. Guarded by {@code this}.
+     * {@code null} if {@link #maxWalArchiveSize} == {@link DataStorageConfiguration#UNLIMITED_WAL_ARCHIVE}.
+     */
+    @Nullable private final Map<Long, Long> segmentSize;
 
-    /** Flag of interrupt waiting on this object. */
-    private volatile boolean interrupted;
+    /**
+     * Segment reservations storage.
+     * {@code null} if {@link #maxWalArchiveSize} == {@link DataStorageConfiguration#UNLIMITED_WAL_ARCHIVE}.
+     */
+    @Nullable private final SegmentReservationStorage reservationStorage;
 
     /**
-     * Adding current WAL archive size in bytes.
+     * Constructor.
      *
-     * @param size Size in bytes.
+     * @param minWalArchiveSize Minimum size of the WAL archive in bytes.
+     * @param maxWalArchiveSize Maximum size of the WAL archive in bytes
+     *      or {@link DataStorageConfiguration#UNLIMITED_WAL_ARCHIVE}.
+     * @param reservationStorage Segment reservations storage.
      */
-    synchronized void addCurrentSize(long size) {
-        curr += size;
+    public SegmentArchiveSizeStorage(
+        long minWalArchiveSize,
+        long maxWalArchiveSize,
+        SegmentReservationStorage reservationStorage
+    ) {
+        this.minWalArchiveSize = minWalArchiveSize;
+        this.maxWalArchiveSize = maxWalArchiveSize;
 
-        if (size > 0)
-            notifyAll();
+        if (maxWalArchiveSize != UNLIMITED_WAL_ARCHIVE) {
+            segmentSize = new TreeMap<>();
+            this.reservationStorage = reservationStorage;
+        }
+        else {
+            segmentSize = null;
+            this.reservationStorage = null;
+        }
     }
 
     /**
-     * Adding reserved WAL archive size in bytes.
-     * Defines a hint to determine if the maximum size is exceeded before a new segment is archived.
+     * Adding the WAL segment size in the archive.
      *
-     * @param size Size in bytes.
+     * @param idx Absolut segment index.
+     * @param curr Current WAL archive size in bytes.
```

Comment:
This javadoc is wrong and name of parameter is bad. Semantics of the parameter is about change of WAL archive size. Lets call it that way: `sizeChange`.